### ISSUE TITLE
NumberFormat/DateTimeFormat getters, with support for named formats.

### ISSUE THREAD

**Issue by skoging** [formatjs/react-intl#1376]
**Is your feature request related to a problem? Please describe.**
There is currently no well supported way to use named formats from react-intl, in combination with `formatToParts()` or `resolvedOptions()` (or any other methods on the format objects other than `format()`).

There have been requests for adding these functions before, but it is reasonable that that is out of the scope of this project.

**Describe the solution you'd like**

```jsx
const intl = useIntl();
const formatter = intl.getNumberFormatter({ format: "currency" }); // <--

// const parts = formatter.formatToParts(100.12);
// const opts = formatter.resolvedOptions();
```

Instead, exporting getter functions for creating the NumberFormat/DateTimeFormat objects with support for named formats is sufficient, and would not introduce breaking changes or add too much noise to the API. 

**Describe alternatives you've considered**
While this is _possible_ to achieve now, without any changes to react-intl, it's not pretty. We've gone through a few iterations on this in our app, but our current implementation boils down to replicating the [internal functionality of react-intl](https://github.com/formatjs/react-intl/blob/master/src/format.ts#L210-L232), to implement these getter functions ourself.

**Additional context**
A central part of our app is forms with a lot of currency and percent fields, and also some date fields. These are formatted and masked, and should ideally be configured the same way as plain number formatting. It should ideally be as simple as 
```jsx
<NumberInput format="currency_input" value={val} onChange={handleChange} />
```
or even
```jsx
function CurrencyInput(props) {
  return <NumberInput format="currency_input" {...props} />;
}
```

This is **not** something that is in scope for react-intl itself, and should be relegated to complimentary packages. 

With this getter feature in place I'd  feel confident enough to open source our formatted input logic, and make it available as complementary packages to react-intl.

**Comment by skoging** [comment]
If this feature is welcome, I'd be open to contributing it myself.

**Comment by longlho** [comment]
I think this already exists (`intl.getNumberFormat()` gives u the instance)

**Comment by skoging** [comment]
if you're thinking of 
```typescript
intl.formatters.getNumberFormat(locale: string, options: Intl.NumberFormatOptions)
````
then yes, that takes the same arguments as the NumberFormatter constructor and returns a memoized instance. However, it does not handle named formats, which is react-intl specific thing. This is handled in the implementation of
```typescript
intl.formatNumber(value: number, options: FormatNumberOptions)
```

The only way I've managed to combine named formats with other formatter methods like formatToParts or resolvedOptions is to reimplement the [logic for getting a named format](https://github.com/formatjs/react-intl/blob/master/src/format.ts#L35-L91). I also implemented the same logic for filtering props, but that's more of a convenience.
